### PR TITLE
feat: add a recovery shell

### PR DESCRIPTION
**Pull request by K900**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -40,6 +40,26 @@ If you want to make NixOS your default distribution, you can do so with
...
+
+```sh
+wsl -d NixOS --system --user root -- /mnt/wslg/distro/bin/nixos-wsl-recovery
+```
+
+This will load the WSL "system" distribution, activate your configuration,
+then chroot into your NixOS system, similar to what `nixos-enter` would do
+on a normal NixOS install.
+
+You can choose an older generation to load with
```
Comment (SuperSandro2000) [inline README.md:55]: Should we add a subcommand to list available generations?

File: modules/wsl-distro.nix [hunk 2]
```diff
@@ -8,6 +8,16 @@ let
     exec ${pkgs.bashInteractive}/bin/sh "$@"
   '';
 
+  nixos-enter' = config.system.build.nixos-enter.overrideAttrs (_: {
```
Comment (SuperSandro2000) [inline modules/wsl-distro.nix:11]: Do we already have the fix for https://github.com/NixOS/nix/issues/7602 ?

File: README.md (hunk 1 above)
Comment (K900) [reply README.md:55]: I don't really want to make that wrapper any bigger than it needs to be, but maybe we should document that.

File: modules/wsl-distro.nix (hunk 2 above)
Comment (K900) [reply modules/wsl-distro.nix:11]: It should be in the next 2.13 release.

**Comment by K900** [comment]
I don't think we should have things in `/nix` that are not managed by Nix itself. It's just asking for trouble later. Otherwise agreed.

File: README.md (hunk 1 above)
Comment (K900) [reply README.md:55]: Once we have it, yes, though I'd like to also have the fallback for people that install from their own tarballs.

**Comment by nzbr** [comment]
Good point. How about `/usr/lib/nixos-wsl` then?

**Comment by K900** [comment]
I'm not sure we want to be touching /usr either... Not really sure what a good place for this is.

**Comment by nzbr** [comment]
image
`/usr` and `/usr/lib` look like this on my installation. I don't see how putting some extra files in there could break anything. 
A great place (but I don't know how feasible it is) would be in the toplevel derivation. That would make the binaries independent of the currently active system so that it'd be possible to just use the one from the previous generation when something breaks the recovery script

**Comment by K900** [comment]
Also, I've changed the shebang and added an explicit error message to detect when the recovery shell is running from inside NixOS, so we can abort.

**Comment by nzbr** [comment]
> we can't actually put this in the toplevel, because the toplevel is an absolute symlink

Unfortunate... It might be worth revisiting when adding a recovery command to the launcher, but for now I'll merge this as-is

**Pull request merged**